File: python/collect_ist_gateway_a.py
#!/usr/bin/env python3
# collect_ist_gateway_a.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:scripts/collect_ist_gateway_a.sh
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import os
import subprocess
import json
import socket
import platform
import urllib.request
from datetime import datetime
from pathlib import Path

def run_command(command, shell=False, default="(nicht ermittelt)"):
    try:
        if shell:
            result = subprocess.run(command, shell=True, capture_output=True, text=True, check=True)
        else:
            result = subprocess.run(command, capture_output=True, text=True, check=True)
        return result.stdout.strip()
    except (subprocess.CalledProcessError, Exception):
        return default

def get_hostname_info():
    try:
        fqdn = socket.getfqdn()
        if fqdn == socket.gethostname():
            fqdn = run_command(["hostname", "-f"], default=socket.gethostname())
    except:
        fqdn = run_command(["hostname", "-f"], default=socket.gethostname())
    return socket.gethostname(), fqdn

def get_public_ip():
    try:
        with urllib.request.urlopen('https://ifconfig.me', timeout=4) as response:
            return response.read().decode('utf-8').strip()
    except:
        return "(nicht ermittelt)"

def get_tailscale_ip():
    return run_command(["tailscale", "ip", "-4"], default="(nicht ermittelt)")

def get_os_info():
    try:
        with open('/etc/os-release', 'r') as f:
            for line in f:
                if line.startswith('PRETTY_NAME='):
                    return line.split('=', 1)[1].strip().strip('"')
    except:
        pass
    return "(nicht ermittelt)"

def get_ipv4_all():
    try:
        result = subprocess.run(["hostname", "-I"], capture_output=True, text=True)
        if result.returncode == 0:
            return ' '.join(result.stdout.strip().split())
        else:
            return "(nicht ermittelt)"
    except:
        return "(nicht ermittelt)"

def check_file_exists(filepath):
    return "vorhanden" if Path(filepath).exists() else "fehlt"

def write_ist_file(ist_file, data):
    with open(ist_file, 'w') as f:
        f.write("# IST-Zustand: Gateway A / Node 1\n\n")
        f.write(f"Stand (lokal): {data['now_local']}  \n")
        f.write(f"Stand (UTC): {data['now_utc']}\n\n")
        f.write("## 1) Identitaet & System\n\n")
        f.write("- Gateway: **A**\n")
        f.write("- Node: **1**\n")
        f.write(f"- Hostname (short): `{data['hostname_short']}`\n")
        f.write(f"- Hostname (FQDN): `{data['hostname_fqdn']}`\n")
        f.write(f"- Architektur: `{data['arch']}`\n")
        f.write(f"- Kernel: `{data['kernel']}`\n")
        f.write(f"- OS: `{data['os_pretty']}`\n")
        f.write(f"- IPv4 (lokal): `{data['ipv4_all']}`\n")
        f.write(f"- Public IPv4: `{data['public_ip']}`\n")
        f.write(f"- Tailscale IPv4: `{data['tailscale_ip']}`\n")
        f.write(f"- OpenClaw Version: `{data['openclaw_ver']}`\n")
        f.write(f"- Node.js Version: `{data['node_ver']}`\n\n")
        f.write("## 2) Arbeitsverzeichnisse\n\n")
        f.write(f"- Basis: `{data['base_dir']}`\n")
        f.write(f"- Funktionell VSCode: `{data['vscode_dir']}`\n")
        f.write(f"- Workspace Doku: `{data['out_dir']}`\n\n")
        f.write("## 3) Kernartefakte (Existenz)\n\n")
        f.write(f"- `{data['openclaw_json']}`: {data['openclaw_json_status']}\n")
        f.write(f"- `{data['env_dot']}`: {data['env_dot_status']}\n")
        f.write(f"- `{data['env_systemd']}`: {data['env_systemd_status']}\n")
        f.write(f"- `{data['base_dir']}/plugins/installs.json`: {data['installs_json_status']}\n")
        f.write(f"- `{data['base_dir']}/plugin-skills`: {data['plugin_skills_status']}\n")

def write_inv_file(inv_file, data):
    with open(inv_file, 'w') as f:
        f.write("# Artefakt-Inventar: Gateway A / Node 1\n\n")
        f.write(f"Stand: {data['now_local']}\n\n")
        f.write("## Top-Level in ~/.openclaw\n\n")
        f.write('```text\n')
        try:
            for item in Path(data['base_dir']).iterdir():
                f.write(f"{item.name}\n")
        except:
            pass
        f.write('```\n\n')
        f.write("## ~/.openclaw/.vscode\n\n")
        f.write('```text\n')
        vscode_path = Path(data['vscode_dir'])
        if vscode_path.exists():
            try:
                for item in vscode_path.iterdir():
                    mode = item.stat().st_mode
                    f.write(f"{'d' if item.is_dir() else '-'} {item.name}\n")
            except:
                f.write("(Zugriff nicht möglich)\n")
        else:
            f.write("(nicht vorhanden)\n")
        f.write('```\n\n')
        f.write("## plugin-skills/\n\n")
        f.write('```text\n')
        plugin_skills_path = Path(data['base_dir']) / "plugin-skills"
        if plugin_skills_path.exists():
            try:
                for item in plugin_skills_path.iterdir():
                    f.write(f"{item.name}\n")
            except:
                f.write("(Zugriff nicht möglich)\n")
        else:
            f.write("(nicht vorhanden)\n")
        f.write('```\n\n')
        f.write("## openclaw.json Backups\n\n")
        f.write('```text\n')
        try:
            backup_files = list(Path(data['base_dir']).glob("openclaw.json.bak*"))
            if backup_files:
                for backup in backup_files:
                    f.write(f"{backup.name}\n")
            else:
                f.write("(keine gefunden)\n")
        except:
            f.write("(keine gefunden)\n")
        f.write('```\n')

def write_cfg_file(cfg_file, data):
    with open(cfg_file, 'w') as f:
        f.write("# OpenClaw Config Snapshot: Gateway A / Node 1\n\n")
        f.write(f"Stand: {data['now_local']}\n\n")
        f.write("## Schluesselpositionen (grep)\n\n")
        f.write('```text\n')
        openclaw_json_path = Path(data['openclaw_json'])
        if openclaw_json_path.exists():
            try:
                with open(openclaw_json_path, 'r') as json_file:
                    lines = json_file.readlines()
                    for i, line in enumerate(lines, 1):
                        if any(key in line for key in ['"gateway"', '"session"', '"dmScope"', '"auth"', '"secrets"', '"tools"', '"plugins"', '"profile"', '"alsoAllow"', '"denyCommands"']):
                            f.write(f"{i}: {line}")
            except:
                f.write("(Fehler beim Lesen)\n")
        else:
            f.write("openclaw.json fehlt\n")
        f.write('```\n\n')
        f.write("## Ausschnitt gateway/session/auth\n\n")
        f.write('```json\n')
        if openclaw_json_path.exists():
            try:
                with open(openclaw_json_path, 'r') as json_file:
                    lines = json_file.readlines()
                    for line in lines[579:780]:
                        f.write(line)
            except:
                f.write('{ "error": "Fehler beim Lesen" }\n')
        else:
            f.write('{ "error": "openclaw.json fehlt" }\n')
        f.write('```\n')

def write_env_file(env_file, data):
    with open(env_file, 'w') as f:
        f.write("# ENV-Status: Gateway A / Node 1\n\n")
        f.write(f"Stand: {data['now_local']}\n\n")
        f.write("## Dateien\n\n")
        f.write('```text\n')
        try:
            env_dot_path = Path(data['env_dot'])
            env_systemd_path = Path(data['env_systemd'])
            if env_dot_path.exists():
                stat_info = env_dot_path.stat()
                f.write(f"-rwx------ 1 user group {stat_info.st_size} {datetime.fromtimestamp(stat_info.st_mtime):%b %d %H:%M} {env_dot_path}\n")
            if env_systemd_path.exists():
                stat_info = env_systemd_path.stat()
                f.write(f"-rwx------ 1 user group {stat_info.st_size} {datetime.fromtimestamp(stat_info.st_mtime):%b %d %H:%M} {env_systemd_path}\n")
        except:
            pass
        f.write('```\n\n')
        f.write("## .env (vollstaendig)\n\n")
        f.write('```dotenv\n')
        env_dot_path = Path(data['env_dot'])
        if env_dot_path.exists():
            try:
                with open(env_dot_path, 'r') as env_file_content:
                    f.write(env_file_content.read())
            except:
                f.write("# Fehler beim Lesen\n")
        else:
            f.write("# .env fehlt\n")
        f.write('```\n\n')
        f.write("## gateway.systemd.env (vollstaendig)\n\n")
        f.write('```dotenv\n')
        env_systemd_path = Path(data['env_systemd'])
        if env_systemd_path.exists():
            try:
                with open(env_systemd_path, 'r') as env_systemd_content:
                    f.write(env_systemd_content.read())
            except:
                f.write("# Fehler beim Lesen\n")
        else:
            f.write("# gateway.systemd.env fehlt\n")
        f.write('```\n')

def write_run_file(run_file, data):
    with open(run_file, 'w') as f:
        f.write("# Laufprotokoll Gateway A / Node 1\n\n")
        f.write(f"- Zeit (lokal): {data['now_local']}\n")
        f.write(f"- Zeit (UTC): {data['now_utc']}\n")
        f.write(f"- Script: {os.path.abspath(__file__)}\n\n")
        f.write("## Erzeugte Dateien\n\n")
        f.write(f"- {os.path.basename(data['ist_file'])}\n")
        f.write(f"- {os.path.basename(data['inv_file'])}\n")
        f.write(f"- {os.path.basename(data['cfg_file'])}\n")
        f.write(f"- {os.path.basename(data['env_file'])}\n")

def main():
    base_dir = os.path.expanduser("~/.openclaw")
    out_dir = os.path.join(base_dir, "workspace", "vscode")
    os.makedirs(out_dir, exist_ok=True)
    
    now_utc = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
    now_local = datetime.now().strftime("%Y-%m-%d %H:%M:%S %Z")
    ts = datetime.now().strftime("%Y%m%d-%H%M%S")
    
    ist_file = os.path.join(out_dir, "IST-ZUSTAND_GATEWAY-A_NODE1.md")
    inv_file = os.path.join(out_dir, "ARTEFAKT-INVENTAR_GATEWAY-A_NODE1.md")
    cfg_file = os.path.join(out_dir, "OPENCLAW-CONFIG-SNAPSHOT_GATEWAY-A_NODE1.md")
    env_file = os.path.join(out_dir, "ENV-STATUS_GATEWAY-A_NODE1.md")
    run_file = os.path.join(out_dir, f"RUN-{ts}.md")
    
    openclaw_json = os.path.join(base_dir, "openclaw.json")
    env_dot = os.path.join(base_dir, ".env")
    env_systemd = os.path.join(base_dir, "gateway.systemd.env")
    vscode_dir = os.path.join(base_dir, ".vscode")
    
    hostname_short, hostname_fqdn = get_hostname_info()
    arch = platform.machine()
    kernel = platform.release()
    os_pretty = get_os_info()
    ipv4_all = get_ipv4_all()
    public_ip = get_public_ip()
    tailscale_ip = get_tailscale_ip()
    openclaw_ver = run_command(["openclaw", "--version"], default="(nicht ermittelt)")
    node_ver = run_command(["node", "-v"], default="(nicht ermittelt)")
    
    data = {
        'now_utc': now_utc,
        'now_local': now_local,
        'hostname_short': hostname_short,
        'hostname_fqdn': hostname_fqdn,
        'arch': arch,
        'kernel': kernel,
        'os_pretty': os_pretty,
        'ipv4_all': ipv4_all,
        'public_ip': public_ip,
        'tailscale_ip': tailscale_ip,
        'openclaw_ver': openclaw_ver,
        'node_ver': node_ver,
        'base_dir': base_dir,
        'vscode_dir': vscode_dir,
        'out_dir': out_dir,
        'openclaw_json': openclaw_json,
        'env_dot': env_dot,
        'env_systemd': env_systemd,
        'ist_file': ist_file,
        'inv_file': inv_file,
        'cfg_file': cfg_file,
        'env_file': env_file
    }
    
    data['openclaw_json_status'] = check_file_exists(openclaw_json)
    data['env_dot_status'] = check_file_exists(env_dot)
    data['env_systemd_status'] = check_file_exists(env_systemd)
    data['installs_json_status'] = check_file_exists(os.path.join(base_dir, "plugins", "installs.json"))
    data['plugin_skills_status'] = "vorhanden" if os.path.isdir(os.path.join(base_dir, "plugin-skills")) else "fehlt"
    
    write_ist_file(ist_file, data)
    write_inv_file(inv_file, data)
    write_cfg_file(cfg_file, data)
    write_env_file(env_file, data)
    write_run_file(run_file, data)
    
    print("OK: IST-Zustand erfasst.")
    try:
        for item in Path(out_dir).iterdir():
            print(f"- {item.name}")
    except:
        pass

if __name__ == "__main__":
    main()

File: javascript/collect_ist_gateway_a.js
#!/usr/bin/env node
// collect_ist_gateway_a.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:scripts/collect_ist_gateway_a.sh
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import fs from 'fs';
import os from 'os';
import path from 'path';
import { execSync } from 'child_process';

const BASE_DIR = path.join(os.homedir(), '.openclaw');
const OUT_DIR = path.join(BASE_DIR, 'workspace', 'vscode');
const NOW_UTC = new Date().toISOString().replace(/\.\d+Z$/, 'Z');
const NOW_LOCAL = new Date().toLocaleString('de-DE') + ' ' + Intl.DateTimeFormat().resolvedOptions().timeZone;
const TS = new Date().toISOString().replace(/[-:]/g, '').replace('T', '-').substring(0, 15);

fs.mkdirSync(OUT_DIR, { recursive: true });

const IST_FILE = path.join(OUT_DIR, 'IST-ZUSTAND_GATEWAY-A_NODE1.md');
const INV_FILE = path.join(OUT_DIR, 'ARTEFAKT-INVENTAR_GATEWAY-A_NODE1.md');
const CFG_FILE = path.join(OUT_DIR, 'OPENCLAW-CONFIG-SNAPSHOT_GATEWAY-A_NODE1.md');
const ENV_FILE = path.join(OUT_DIR, 'ENV-STATUS_GATEWAY-A_NODE1.md');
const RUN_FILE = path.join(OUT_DIR, `RUN-${TS}.md`);

const OPENCLAW_JSON = path.join(BASE_DIR, 'openclaw.json');
const ENV_DOT = path.join(BASE_DIR, '.env');
const ENV_SYSTEMD = path.join(BASE_DIR, 'gateway.systemd.env');
const VSCODE_DIR = path.join(BASE_DIR, '.vscode');

let HOSTNAME_FQDN = '';
try {
  HOSTNAME_FQDN = os.hostname();
} catch (e) {
  HOSTNAME_FQDN = os.hostname();
}

const HOSTNAME_SHORT = os.hostname();
const ARCH = process.arch;
const KERNEL = os.release();
let OS_PRETTY = 'Unbekannt';
try {
  const osRelease = fs.readFileSync('/etc/os-release', 'utf8');
  const match = osRelease.match(/^PRETTY_NAME=(.*)$/m);
  if (match) {
    OS_PRETTY = match[1].replace(/"/g, '');
  }
} catch (e) {
  // leer
}

const IPV4_ALL = Object.values(os.networkInterfaces())
  .flat()
  .filter(iface => iface.family === 'IPv4' && !iface.internal)
  .map(iface => iface.address)
  .join(' ');

let PUBLIC_IP = '';
try {
  PUBLIC_IP = execSync('curl -4 -s --max-time 4 ifconfig.me', { encoding: 'utf8' }).trim();
} catch (e) {
  PUBLIC_IP = '(nicht ermittelt)';
}

let TAILSCALE_IP = '';
try {
  TAILSCALE_IP = execSync('tailscale ip -4', { encoding: 'utf8' }).split('\n')[0].trim();
} catch (e) {
  TAILSCALE_IP = '(nicht ermittelt)';
}

let OPENCLAW_VER = '';
try {
  OPENCLAW_VER = execSync('openclaw --version', { encoding: 'utf8' }).trim();
} catch (e) {
  OPENCLAW_VER = '(nicht ermittelt)';
}

let NODE_VER = process.version;

if (!PUBLIC_IP) PUBLIC_IP = '(nicht ermittelt)';
if (!TAILSCALE_IP) TAILSCALE_IP = '(nicht ermittelt)';
if (!OPENCLAW_VER) OPENCLAW_VER = '(nicht ermittelt)';
if (!NODE_VER) NODE_VER = '(nicht ermittelt)';

const istContent = `# IST-Zustand: Gateway A / Node 1

Stand (lokal): ${NOW_LOCAL}  
Stand (UTC): ${NOW_UTC}

## 1) Identitaet & System

- Gateway: **A**
- Node: **1**
- Hostname (short): \`${HOSTNAME_SHORT}\`
- Hostname (FQDN): \`${HOSTNAME_FQDN}\`
- Architektur: \`${ARCH}\`
- Kernel: \`${KERNEL}\`
- OS: \`${OS_PRETTY}\`
- IPv4 (lokal): \`${IPV4_ALL}\`
- Public IPv4: \`${PUBLIC_IP}\`
- Tailscale IPv4: \`${TAILSCALE_IP}\`
- OpenClaw Version: \`${OPENCLAW_VER}\`
- Node.js Version: \`${NODE_VER}\`

## 2) Arbeitsverzeichnisse

- Basis: \`${BASE_DIR}\`
- Funktionell VSCode: \`${VSCODE_DIR}\`
- Workspace Doku: \`${OUT_DIR}\`

## 3) Kernartefakte (Existenz)

- \`${OPENCLAW_JSON}\`: ${fs.existsSync(OPENCLAW_JSON) ? 'vorhanden' : 'fehlt'}
- \`${ENV_DOT}\`: ${fs.existsSync(ENV_DOT) ? 'vorhanden' : 'fehlt'}
- \`${ENV_SYSTEMD}\`: ${fs.existsSync(ENV_SYSTEMD) ? 'vorhanden' : 'fehlt'}
- \`${path.join(BASE_DIR, 'plugins/installs.json')}\`: ${fs.existsSync(path.join(BASE_DIR, 'plugins/installs.json')) ? 'vorhanden' : 'fehlt'}
- \`${path.join(BASE_DIR, 'plugin-skills')}\`: ${fs.existsSync(path.join(BASE_DIR, 'plugin-skills')) ? 'vorhanden' : 'fehlt'}
`;

fs.writeFileSync(IST_FILE, istContent);

let invContent = `# Artefakt-Inventar: Gateway A / Node 1

Stand: ${NOW_LOCAL}

## Top-Level in ~/.openclaw

\`\`\`text
`;
try {
  const files = fs.readdirSync(BASE_DIR);
  invContent += files.join('\n');
} catch (e) {
  // leer
}
invContent += `
\`\`\`

## ~/.openclaw/.vscode

\`\`\`text
`;
if (fs.existsSync(VSCODE_DIR)) {
  try {
    const files = fs.readdirSync(VSCODE_DIR);
    invContent += files.map(f => {
      const stat = fs.statSync(path.join(VSCODE_DIR, f));
      return `${stat.isDirectory() ? 'd' : '-'} ${f}`;
    }).join('\n');
  } catch (e) {
    invContent += "(Fehler beim Lesen)";
  }
} else {
  invContent += "(nicht vorhanden)";
}
invContent += `
\`\`\`

## plugin-skills/

\`\`\`text
`;
if (fs.existsSync(path.join(BASE_DIR, 'plugin-skills'))) {
  try {
    const files = fs.readdirSync(path.join(BASE_DIR, 'plugin-skills'));
    invContent += files.join('\n');
  } catch (e) {
    // leer
  }
} else {
  invContent += "(nicht vorhanden)";
}
invContent += `
\`\`\`

## openclaw.json Backups

\`\`\`text
`;
try {
  const files = fs.readdirSync(BASE_DIR).filter(f => f.startsWith('openclaw.json.bak'));
  if (files.length > 0) {
    invContent += files.join('\n');
  } else {
    invContent += "(keine gefunden)";
  }
} catch (e) {
  invContent += "(keine gefunden)";
}
invContent += `
\`\`\`
`;

fs.writeFileSync(INV_FILE, invContent);

let cfgContent = `# OpenClaw Config Snapshot: Gateway A / Node 1

Stand: ${NOW_LOCAL}

## Schluesselpositionen (grep)

\`\`\`text
`;
if (fs.existsSync(OPENCLAW_JSON)) {
  try {
    const content = fs.readFileSync(OPENCLAW_JSON, 'utf8');
    const lines = content.split('\n');
    lines.forEach((line, index) => {
      if (line.match(/"gateway"|\"session\"|\"dmScope\"|\"auth\"|\"secrets\"|\"tools\"|\"plugins\"|\"profile\"|\"alsoAllow\"|\"denyCommands\"/)) {
        cfgContent += `${index + 1}: ${line}\n`;
      }
    });
  } catch (e) {
    // leer
  }
} else {
  cfgContent += "openclaw.json fehlt\n";
}
cfgContent += `\`\`\`

## Ausschnitt gateway/session/auth

\`\`\`json
`;
if (fs.existsSync(OPENCLAW_JSON)) {
  try {
    const content = fs.readFileSync(OPENCLAW_JSON, 'utf8');
    const lines = content.split('\n').slice(579, 780);
    cfgContent += lines.join('\n');
  } catch (e) {
    cfgContent += '{ "error": "Fehler beim Lesen" }';
  }
} else {
  cfgContent += '{ "error": "openclaw.json fehlt" }';
}
cfgContent += `
\`\`\`
`;

fs.writeFileSync(CFG_FILE, cfgContent);

let envContent = `# ENV-Status: Gateway A / Node 1

Stand: ${NOW_LOCAL}

## Dateien

\`\`\`text
`;
try {
  const stats1 = fs.statSync(ENV_DOT);
  const stats2 = fs.statSync(ENV_SYSTEMD);
  envContent += `${stats1.mode} ${ENV_DOT}\n`;
  envContent += `${stats2.mode} ${ENV_SYSTEMD}\n`;
} catch (e) {
  // leer
}
envContent += `\`\`\`

## .env (vollstaendig)

\`\`\`dotenv
`;
if (fs.existsSync(ENV_DOT)) {
  try {
    envContent += fs.readFileSync(ENV_DOT, 'utf8');
  } catch (e) {
    envContent += "# Fehler beim Lesen";
  }
} else {
  envContent += "# .env fehlt";
}
envContent += `
\`\`\`

## gateway.systemd.env (vollstaendig)

\`\`\`dotenv
`;
if (fs.existsSync(ENV_SYSTEMD)) {
  try {
    envContent += fs.readFileSync(ENV_SYSTEMD, 'utf8');
  } catch (e) {
    envContent += "# Fehler beim Lesen";
  }
} else {
  envContent += "# gateway.systemd.env fehlt";
}
envContent += `
\`\`\`
`;

fs.writeFileSync(ENV_FILE, envContent);

const runContent = `# Laufprotokoll Gateway A / Node 1

- Zeit (lokal): ${NOW_LOCAL}
- Zeit (UTC): ${NOW_UTC}
- Script: ${process.argv[1]}

## Erzeugte Dateien

- ${path.basename(IST_FILE)}
- ${path.basename(INV_FILE)}
- ${path.basename(CFG_FILE)}
- ${path.basename(ENV_FILE)}
`;

fs.writeFileSync(RUN_FILE, runContent);

console.log("OK: IST-Zustand erfasst.");
try {
  const files = fs.readdirSync(OUT_DIR);
  files.forEach(file => console.log(`- ${file}`));
} catch (e) {
  // leer
}
